alerts/config.go

```diff
@@ -5,116 +5,179 @@ package alerts
-	dur := d.Delay.Duration()
-	validator.Duration("delay", &dur).GreaterThan(0 * time.Minute)
+func validateRepeat(value time.Duration, errorReporter structure.ErrorReporter) {
+	if value == 0 {
+		return
+	}
+	if value < RepeatMin {
+		errorReporter.ReportError(validator.ErrorValueNotGreaterThanOrEqualTo(value, RepeatMin))
+	}
+	if value > RepeatMax {
+		errorReporter.ReportError(validator.ErrorValueNotLessThanOrEqualTo(value, RepeatMax))
+	}
 }
 
-// ThresholdMixin adds a configurable threshold.
-type ThresholdMixin struct {
+// UrgentLowAlert extends Base with a threshold.
+type UrgentLowAlert struct {
+	Base `bson:",inline"`
 	// Threshold is compared the current value to determine if an alert should
 	// be triggered.
 	Threshold `json:"threshold"`
 }
 
-func (t ThresholdMixin) Validate(validator structure.Validator) {
-	t.Threshold.Validate(validator)
+func (d UrgentLowAlert) Validate(validator structure.Validator) {
+	d.Base.Validate(validator)
+	d.Threshold.Validate(validator)
+	validateThresholdInNativeUnits(validator, d.Threshold, 40, 55, glucose.MgdL)
 }
 
-// WithThreshold extends Base with ThresholdMixin.
-type WithThreshold struct {
-	Base           `bson:",inline"`
-	ThresholdMixin `bson:",inline"`
+// validateThresholdInNativeUnits formats validation errors in the same units
+// submitted by the user.
+//
+// This provides easier to understand validation error messages.
+func validateThresholdInNativeUnits(validator structure.Validator, threshold Threshold, min, max float64, units string) {
+	nativeMin := glucose.Convert(min, glucose.MgdL, threshold.Units)
+	nativeMax := glucose.Convert(max, glucose.MgdL, threshold.Units)
+	validator.Float64("threshold.value", &threshold.Value).InRange(nativeMin, nativeMax)
+}
+
+// NotLoopingAlert extends Base with a delay.
+type NotLoopingAlert struct {
+	Base  `bson:",inline"`
+	Delay DurationMinutes `json:"delay,omitempty"`
 }
 
-func (d WithThreshold) Validate(validator structure.Validator) {
+func (d NotLoopingAlert) Validate(validator structure.Validator) {
 	d.Base.Validate(validator)
-	d.ThresholdMixin.Validate(validator)
+	dur := d.Delay.Duration()
+	validator.Duration("delay", &dur).InRange(0, 2*time.Hour)
 }
 
-// WithDelay extends Base with DelayMixin.
-type WithDelay struct {
-	Base       `bson:",inline"`
-	DelayMixin `bson:",inline"`
+// NoCommunicationAlert extends Base with a delay.
+type NoCommunicationAlert struct {
+	Base  `bson:",inline"`
+	Delay DurationMinutes `json:"delay,omitempty"`
+}
+
+func (d NoCommunicationAlert) Validate(validator structure.Validator) {
+	d.Base.Validate(validator)
+	dur := d.Delay.Duration()
+	validator.Duration("delay", &dur).InRange(0, 6*time.Hour)
+}
+
+// LowAlert extends Base with threshold and a delay.
+type LowAlert struct {
+	Base `bson:",inline"`
+	// Threshold is compared the current value to determine if an alert should
+	// be triggered.
+	Threshold `json:"threshold"`
+	Delay     DurationMinutes `json:"delay,omitempty"`
```

Comment:
Not sure if this got turned off or is no longer used, but is the code formatter working? IIRC, I think the struct tags should be aligned here. Perhaps it is no longer a thing.